modules/frontend/frontend.go

```diff
@@ -25,15 +25,107 @@ import (
 	"github.com/grafana/tempo/pkg/util"
 )
 
-// NewTripperware returns a Tripperware configured with a middleware to split requests
+// NewTripperware returns a Tripperware configured with a middleware to route, split and dedupe requests.
 func NewTripperware(cfg Config, logger log.Logger, registerer prometheus.Registerer) (queryrange.Tripperware, error) {
-	level.Info(logger).Log("msg", "creating tripperware in query frontend to shard queries")
+	level.Info(logger).Log("msg", "creating tripperware in query frontend")
+
+	tracesTripperware := NewTracesTripperware(cfg, logger)
+	searchTripperware := NewSearchTripperware(logger)
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		traces := tracesTripperware(next)
+		search := searchTripperware(next)
+
+		return newFrontendRoundTripper(next, traces, search, logger, registerer)
+	}, nil
+}
+
+type frontendRoundTripper struct {
+	next, traces, search http.RoundTripper
+	logger               log.Logger
+	queriesPerTenant     *prometheus.CounterVec
+}
+
+func newFrontendRoundTripper(next, traces, search http.RoundTripper, logger log.Logger, registerer prometheus.Registerer) frontendRoundTripper {
 	queriesPerTenant := promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
 		Namespace: "tempo",
 		Name:      "query_frontend_queries_total",
 		Help:      "Total queries received per tenant.",
 	}, []string{"tenant"})
 
+	return frontendRoundTripper{
+		next:             next,
+		traces:           traces,
+		search:           search,
+		logger:           logger,
+		queriesPerTenant: queriesPerTenant,
+	}
+}
+
+func (r frontendRoundTripper) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	start := time.Now()
+	// tracing instrumentation
+	span, ctx := opentracing.StartSpanFromContext(req.Context(), "frontend.Tripperware")
+	defer span.Finish()
+
+	orgID, _ := user.ExtractOrgID(req.Context())
+	r.queriesPerTenant.WithLabelValues(orgID).Inc()
+	span.SetTag("orgID", orgID)
+
+	// for context propagation with traceID set
+	req = req.WithContext(ctx)
+
+	// route the request to the appropriate RoundTripper
+	switch op := getOperation(req.URL.Path); op {
+	case TracesOp:
+		resp, err = r.traces.RoundTrip(req)
+	case SearchOp:
+		resp, err = r.search.RoundTrip(req)
+	default:
+		resp, err = r.next.RoundTrip(req)
```

Comment:
This is a bit convoluted: in `NewTripperware` (at the top of the file) we are wrapping `next` with traces and search tripperware. So if we call `r.traces.RoundTrip` this will execute the traces stuff and then call `next.RoundTrip`.
The default case skips the traces and search tripperware and directly calls `next.RoundTrip`.
So all calls end up executing `next.RoundTrip`, but when the path matches `/api/traces` or `/api/search` we perform extra validation and conversions.

This is basically the same code [as Loki](https://github.com/grafana/loki/blob/17f4a73a24c8afc36bcb4e2795c6b1e06d4b95a1/pkg/querier/queryrange/roundtrip.go#L144) btw but a lot simpler.

Reply:
Per offline discussion, we agreed that the "default case" should actually never be called. I will be adding a warning log & a comment here.